### DISCUSSION TITLE
How does the computation of closed loop mechanism work?

### DISCUSSION THREAD

**Discussion by vicltz** [isaac-sim/IsaacLab#1501] [General]
The [Articulation API](https://docs.omniverse.nvidia.com/extensions/latest/ext_physics/articulations.html#closing-loops) allows to model closed loop mechanism / parallel mechanism in Isaac Sim. Some topics advise to lower the simulation step in order to make it work well.

However, there is no information in the documentation that explain the computation under the hood or any information in order to make it work well.

I would like to develop a learning environnement with robots having parallel linkages.

This raises the following questions:
 
1. How does the simulation work when using this API ? For instance, in the MuJoCo simulator, the dynamics computation is solved as an optimization problem with closed loops modeled as soft constraints as explained in [[1]](https://mujoco.readthedocs.io/en/stable/modeling.html#solver-parameters)[[2]](https://mujoco.readthedocs.io/en/stable/computation/index.html#constraint-solver). Where constraints error is turned into an acceleration at loop closure. Does Isaac work in a similar way ?

2. Is there any general advice to obtain the best behavior from this API ?

3. Is this feature mature enough for robot learning, i.e Reinforcement Learning and Sim2Real ? If yes, does it scale to many robots at the same time with decent compute time ?

 4. Does anyone have ever used it for reinforcement learning for humanoid robots? For instance, we can see on the [project GROOT](https://developer.nvidia.com/project-gr00t) page a video with the Agility Digit, Fourier Intelligence GR-1 and Apptronick Apollo robots, that all have closed kinematic chains in their architecture , evolving in simulation . Was the Articulation API used for this ?

Thanks.

**Comment by StrainFlow** [comment]
1. Articulations use a reduced coordinate solver which is very accurate but cannot handle closed loops. When you create a loop and then choose one joint to be excluded from the articulation, I'm not sure how it handles that specific joint but it is noticeably less accurate in the simulation. 

2. The best advice I can give is to play with the solver settings and pick a joint to exclude that can be modeled with lower accuracy. If the joint isn't converging well there will be noticeable movement/jitters/issues with that joint. 

3. Absolutely. I've been able to get models that work pretty well even cutting joints out of articulation loops. It's just a matter of your assumptions. You might need to introduce additional domain randomization to take the sim2real gap into account. I haven't scaled to a large number of robots in an environment, but that last joint is modeled with a method that uses much less computation than the reduced coordinate method, so I would be surprised if that was an issue. 

4. I haven't seen this as an issue with humanoid robots because most of them don't have closed loops in their articulations, so the reduced coordinate solver doesn't have an issue with them. 

Depending on what you're doing, you might be able to use an approximate model that eliminates any loops and models the system more accurately. I was working with an ackermann robot and having a lot of trouble with the articulations because they were closed-loop and they had a lot of spherical bearings, so the kinematic motion was complex. I tweaked the suspension gemeometry so that while it was different from the real car's geometry, I could reduce all spherical joints, simplify the kinematics, and get an overall better simulation.

**Comment by vicltz** [comment]
Thanks a lot for your answer. Let me react to each of your points to clarify a few things.

1. If I understand you well, what is called "Articulation" in Isaac is the API feature that allows simulation in minimal (Lagrangian) coordinates, whereas the usual "joints" are simulated with maximal coordinates (x,y,z,angles) for each rigid body. I feel like it says nothing about how the `Exclude from Articulation` flag work ? Since the documentation does not say much, can you please elaborate on this point ?

3. Not sure what you mean by " that last joint is modeled with a method that uses much less computation than the reduced coordinate method" ? Can you explain ?

4. Actually, the ones that can be seen in the project [GROOT video](https://developer.nvidia.com/project-gr00t) , which are Agility Digit, Fourier Intelligence GR1 and Apptronick Apollo all have closed kinematic chain in their design (in their ankle and knees). By looking at the videos, it looks like these mechanism are simulated. But they could also use a simplified serial model and what we see on the video could be just visual meshes.

**Comment by vicltz** [comment]
UP
